### ISSUE TITLE
while using NanGuardMode(nan_is_error=True ) -->  basic.py  in min raise NotImplementedError() NotImplementedError

### ISSUE THREAD

**Issue by jc-Rosichini** [Theano/Theano#5701]
Just went though the following issue when trying to catch a Nan using mode=NanGuardMode(nan_is_error=True ) 

**my config:**
theano: 0.9.0rc2.dev-674f8d5148cb8dcea446e7243e629d8f83647da5
numpy: 1.11.0
pygpu: 0.6.1
lasagne: 0.2.dev1
Using cuDNN version 5105
GeForce GTX 980M
Windows 10 x64

**compiled training function**
train_fn = theano.function([net.input_var, net.target_var], [loss, I, U, acc], updates=updates
 , mode=NanGuardMode(nan_is_error=True )                               

> Compilation starts at 2017-03-12 16:12:29
optimizer: rmsprop, initial learning rate:  0.001
train compilation took 169.327 seconds
Compilation starts at 2017-03-12 16:15:22
valid compilation took 116.101 seconds
Validation score before pursuing training 0.0539679
Training restarts at 2017-03-12 16:17:24

> Traceback (most recent call last):
>   File "C:\Users\PCGP1\Anaconda3\lib\site-packages\theano\gof\vm.py", line 482, in __call__
>     _, dt = self.run_thunk_of_node(current_apply)
>   File "C:\Users\PCGP1\Anaconda3\lib\site-packages\theano\gof\vm.py", line 402, in run_thunk_of_node
>     compute_map=self.compute_map,
>   File "C:\Users\PCGP1\Anaconda3\lib\site-packages\theano\compile\nanguardmode.py", line 344, in nan_check
>     do_check_on(storage_map[var][0], node)
>   File "C:\Users\PCGP1\Anaconda3\lib\site-packages\theano\compile\nanguardmode.py", line 296, in do_check_on
>     if contains_nan(value, nd, var):
>   File "C:\Users\PCGP1\Anaconda3\lib\site-packages\theano\compile\nanguardmode.py", line 121, in contains_nan
>     return np.isnan(f_gpua_min(arr.reshape(arr.size)))
>   File "C:\Users\PCGP1\Anaconda3\lib\site-packages\theano\compile\nanguardmode.py", line 227, in result
>     f = theano.function([guard_in], op(guard_in),
>   File "C:\Users\PCGP1\Anaconda3\lib\site-packages\theano\tensor\basic.py", line 1742, in min
>     raise NotImplementedError()
> NotImplementedError

**Comment by lamblin** [comment]
In fact, the problem is that we are trying to compute the minimum of an unsigned GpuArray, which is not implemented.
However, this only happens when trying to detect NaN in that array, which cannot happen for unsigned ints.

The solution would be to update `contains_nan` and `contains_inf` to skip checks for discrete-type values.

**Comment by nouiz** [comment]
It is now fixed in the master, so closing. But I'll still commit a PR that skip the check for int.

**Comment by nouiz** [comment]
https://github.com/Theano/Theano/pull/6025 does this speed up. As this is fixed in master, I close now.

**Comment by lamblin** [comment]
The issue is definitely still there in the master. With the following script:
```python
import theano
from theano import tensor
from theano.compile.nanguardmode import NanGuardMode


x = tensor.vector(dtype='uint64')
y = x + 1
f = theano.function([x], y, mode=NanGuardMode(nan_is_error=True))
print(f([23, 7]))
```
I have:
```
THEANO_FLAGS=device=cuda0 python test_5701.py
Using cuDNN version 5105 on context None
Mapped name None to device cuda0: GeForce GTX 750 Ti (0000:01:00.0)
Traceback (most recent call last):
  File "test_5701.py", line 9, in <module>
    print(f([23, 7]))
  File "/u/lamblinp/code/Theano/theano/compile/function_module.py", line 884, in __call__
    self.fn() if output_subset is None else\
  File "/u/lamblinp/code/Theano/theano/gof/vm.py", line 519, in __call__
    storage_map=storage_map)
  File "/u/lamblinp/code/Theano/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/u/lamblinp/code/Theano/theano/gof/vm.py", line 486, in __call__
    _, dt = self.run_thunk_of_node(current_apply)
  File "/u/lamblinp/code/Theano/theano/gof/vm.py", line 404, in run_thunk_of_node
    compute_map=self.compute_map,
  File "/u/lamblinp/code/Theano/theano/compile/nanguardmode.py", line 269, in nan_check
    do_check_on(storage_map[var][0], node)
  File "/u/lamblinp/code/Theano/theano/compile/nanguardmode.py", line 224, in do_check_on
    if contains_nan(value, nd, var):
  File "/u/lamblinp/code/Theano/theano/compile/nanguardmode.py", line 110, in contains_nan
    return np.isnan(f_gpua_min(arr.reshape(arr.size)))
  File "/u/lamblinp/code/Theano/theano/compile/nanguardmode.py", line 158, in result
    f = theano.function([guard_in], op(guard_in),
  File "/u/lamblinp/code/Theano/theano/tensor/basic.py", line 1720, in min
    raise NotImplementedError()
NotImplementedError: 
Apply node that caused the error: GpuFromHost<None>(<TensorType(uint64, vector)>)
Toposort index: 0
Inputs types: [TensorType(uint64, vector)]
Inputs shapes: [(2,)]
Inputs strides: [(8,)]
Inputs values: [array([23,  7], dtype=uint64)]
Outputs clients: [[GpuElemwise{add,no_inplace}(GpuArrayConstant{[ 1.]}, GpuFromHost<None>.0)]]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_compile'. If that does not work, Theano optimizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.
zsh: exit 1     THEANO_FLAGS=device=cuda0 python test_5701.py
```

**Comment by lamblin** [comment]
Fixed now that #6025 has been merged.
